File: models/Bullet.ts
import { degreeToRad, rotatePoint } from "../helpers/math";
import Position from "../interfaces/Position";
import Damager from "../interfaces/Damager";
import Damageable from "../interfaces/Damageable";
import Vehicle from './Vehicle';

class Bullet extends Damager {
  private x: number = 0;
  private y: number = 0;
  private speed: number = 0;
  private angle: number = 90;
  private angle_rad: number = 0;
  private vehicle_id: string = '?';

  static RADIUS = 2;

  constructor (
    x: number = 0,
    y: number = 0,
    angle: number = 0,
    vehicle_id: string = '?',
    speed: number = 10,
  ) {
    super()
    this.x = x;
    this.y = y;
    this.speed = speed;
    this.angle = angle;
    this.angle_rad = degreeToRad(angle);
    this.vehicle_id = vehicle_id;
  }

  getPosition(): Position {
    return { x: this.x, y: this.y };
  }

  getSpeed(): number { return this.speed; }
  getAngle(): number { return this.angle; }
  getAngleRad(): number { return this.angle_rad; }
  getVehicleId(): string { return this.vehicle_id; }

  update () {
    this.y = this.y + (Math.sin(this.angle_rad) * this.speed);
    this.x = this.x + (Math.cos(this.angle_rad) * this.speed);
  }

  getBoundaries(): Array<Position> {
    return [
      { x: this.x, y: this.y - Bullet.RADIUS },
      { x: this.x + Bullet.RADIUS, y: this.y },
      { x: this.x, y: this.y + Bullet.RADIUS },
      { x: this.x - Bullet.RADIUS, y: this.y },
    ];
  }

  getDamageToApply(damageable: Damageable): number {
    if (damageable instanceof Vehicle) {
      if (damageable.getVehicleId() == this.vehicle_id) return 0

      return 30
    }
    return 2
  };
  
}

export default Bullet;

File: rust/models/Bullet.rs
use crate::interfaces::{Position, Damager, Damageable};
use crate::helpers::math::{degree_to_rad};

pub struct Bullet {
    x: f64,
    y: f64,
    speed: f64,
    angle: f64,
    angle_rad: f64,
    vehicle_id: String,
}

impl Bullet {
    pub const RADIUS: f64 = 2.0;

    pub fn new(x: f64, y: f64, angle: f64, vehicle_id: String, speed: f64) -> Self {
        let angle_rad = degree_to_rad(angle);
        Bullet { x, y, speed, angle, angle_rad, vehicle_id }
    }

    pub fn get_position(&self) -> Position {
        Position { x: self.x, y: self.y }
    }

    pub fn get_speed(&self) -> f64 { self.speed }
    pub fn get_angle(&self) -> f64 { self.angle }
    pub fn get_angle_rad(&self) -> f64 { self.angle_rad }
    pub fn get_vehicle_id(&self) -> &String { &self.vehicle_id }

    pub fn update(&mut self) {
        self.y += self.angle_rad.sin() * self.speed;
        self.x += self.angle_rad.cos() * self.speed;
    }

    pub fn get_boundaries(&self) -> Vec<Position> {
        vec![
            Position { x: self.x, y: self.y - Bullet::RADIUS },
            Position { x: self.x + Bullet::RADIUS, y: self.y },
            Position { x: self.x, y: self.y + Bullet::RADIUS },
            Position { x: self.x - Bullet::RADIUS, y: self.y },
        ]
    }

    pub fn get_damage_to_apply(&self, damageable: &Damageable) -> f64 {
        if let Some(vehicle) = damageable.as_vehicle() {
            if vehicle.get_vehicle_id() == self.vehicle_id {
                return 0.0;
            }
            return 30.0;
        }
        2.0
    }
}
